src/QirRuntime/lib/Tracer/tracer.hpp

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <limits>
+#include <memory>
+#include <ostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "CoreTypes.hpp"
+#include "QuantumApi_I.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    using OpId = int;
+    using Time = int;
+    using Duration = int;
+    using LayerId = size_t;
+
+    constexpr LayerId INVALID = std::numeric_limits<size_t>::max();
+
+    /*==================================================================================================================
+        Layer
+    ==================================================================================================================*/
+    struct Layer
+    {
+        // Width of the layer on the time axis.
+        const Duration duration;
+
+        // Start time of the layer.
+        const Time startTime;
+
+        // Quantum operations, assigned to this layer.
+        std::unordered_map<OpId, int /*count of the op with this id*/> operations;
+
+        // Optional id, if the layer represents a global barrier.
+        OpId barrierId = -1;
+
+        Layer(Duration duration, Time startTime)

```

Comment:
Everywhere in the ReadMe.md you refer to the layer ("L(T, N)") such that the time goes first, and then the duration.
Here in the constructor - vice versa. 
The user of this code can easily make a mistake. I recommend the same order everywhere.

Reply:
Totally agree. Not sure why my brain flipped to the reverse order when implementing layering...